indexer/ai/src/main/java/com/google/cloudsearch/ai/BaseAISkill.java

```diff
@@ -0,0 +1,133 @@
+package com.google.cloudsearch.ai;
+
+import com.google.cloudsearch.exceptions.InvalidConfigException;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import java.util.*;
+
+/**
+ * Abstract class BaseAISkill implements the common functionality required for the different skills.
+ */
+public abstract class BaseAISkill implements AISkill {
+
+    private String aiSkillName = "";
+    private List<OutputMapping> outputMappings = new ArrayList<OutputMapping>();
+    private Map<String, Boolean> schemaInfo = new HashMap<String, Boolean>();
+    private Boolean isSchemaInfoStored = false;
+
+    /**
+     * Parses the CloudsSearch Schema and stores the object and property names
+     * @param schema    CloudSearch schema
+     */
+    private void storeSchemaInfo(JSONObject schema){
```

Comment:
We should parse the schema just once right and not again for every skill? 
This info can be passed in the constructor of the Skills class?

Reply:
It is parsed once, I have a boolean variable for the check. I had the constructor then I had removed it and changed the structure a bit.

Reply:
Okay I checked the workflow, it is actually called from the constructor of each skill. And since each. skill have different data structures, it is stored separately for each.